Native Typed Persistence Behavior Implementation - but more Immutable

ktoso opened the pull request:
I'm not convienced that this style makes the implementation any simpler, but it works. Likely some bugs somewhere still, but could be an initial merge... (Same as the other PR)

This mirrors https://github.com/akka/akka/pull/24454 but in an "more immutable" style.

ktoso committed:
=per native typed implementation of Eventsourced=>PersistendBehavior
timeout is explicitly a message of Command

persitAll and chainable side effects work well

more tests pasing

additional sanity check that mutable behaviors work as expected

unstashing needs to "loop through" the AdapterActor otherwise Stopped
won't work

solve unstashing/stop issue, by not randomly init()ing, but unstashing

snapshotting works

all tests green

rebased

nicer log source

remove IncomingCommand wrapper, we dont need it

no need for shared counter

remove not needed methods and state

more state cleanup, using Behaviors.same

reminder that we DO need that same alias, since stash does not work with
the Behavior.same

introduce config for stash buffer

stopping now works after persisting

compile fix

cleanup

reduced number of adapter styles needed for co-existence of persistence

final cleanup done, less passing around 40 objects, carriers provided
now

jrudolph committed:
separate PersistentBehavior API and implementation

jrudolph committed:
Allow internal custom implementations of DeferredBehavior

Pass around context for stashing

jrudolph committed:
Rename PersistentBehaviorImpl => EventsourcedSetup and pass setup around directly

ktoso committed:
WIP towards immutable style

compiles, does not work..

patriknw committed:
reduce boilerplate for passing setup (and [C,E,S])

use an enclosing class that holds the setup

patriknw committed:
EventsourcedRunning, internalPersist, internalPersistAll

* I don't think the JournalInteractions should be conserned with
  next behavior

patriknw committed:
make PersistingEvent MutableBehavior

fix some tests, snapshots still not working

ktoso committed:
add MDC logging to StdOutLogger, making withMdc in Akka typed usable

ktoso committed:
may want to benchmark typed persistence

ktoso committed:
cleanup and make snapshots work, make event timeout work too

sadly does not work without MODULE$ did a ticket for it

ktoso commented:
Aims to resolve #23694

patriknw commented on akka-persistence-typed/src/test/java/akka/persistence/typed/javadsl/PersistentActorTest.java:
```diff
@@ -26,7 +26,7 @@
 
   @ClassRule
-  public static final TestKitJunitResource testKit = new TestKitJunitResource(PersistentActorSpec$.MODULE$.config());
+  public static final TestKitJunitResource testKit = new TestKitJunitResource(PersistentBehaviorSpec$.MODULE$.config());
```
boy scout: add some accessor instead of using MODULE

ktoso replied:
Running into 

```
error] /Users/ktoso/code/akka/akka-persistence-typed/src/test/java/akka/persistence/typed/javadsl/PersistentActorTest.java:29:1: non-static method config() cannot be referenced from a static context
[error]   public static final TestKitJunitResource testKit = new TestKitJunitResource(PersistentBehaviorSpec.config());
```

ktoso replied:
Made a ticket https://github.com/akka/akka/issues/24692

patriknw commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala:
```diff
@@ -3,7 +3,7 @@
 package akka.stream.impl.io
 
-import java.io.{ IOException, InputStream }
+import java.io.{ BufferedOutputStream, ByteArrayOutputStream, IOException, InputStream }
```
?

patriknw commented on akka-persistence/src/main/scala/akka/persistence/RecoveryPermitter.scala:
```diff
@@ -18,9 +18,11 @@ import akka.actor.Terminated
-  case object RecoveryPermitGranted
-  case object ReturnRecoveryPermit
+  sealed trait Protocol
+  sealed trait Reply extends Protocol
```
then there should also be a `Request`?

ktoso replied:
true

patriknw reviewed (commented):
First batch of comments from detailed review. I still have EventsourcedRunning and EventsourcedRecoveringEvents remaining.

There are also some "match may not be exhaustive" warnings that should be silenced.

ktoso commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedSetup.scala:
```diff
@@ -0,0 +1,67 @@
+
+  settings: EventsourcedSettings,
+
+  internalStash: StashBuffer[InternalProtocol] // FIXME would be nice here... but stash is mutable :\\\\\\\
```
the entire setup is immutable, yet it carries the mutable internalStash; felt that's a bit weird, but perhaps it's fine and removing the fixme is ok; Alternative would be to carry the mutable state explicitly, but I don't think it's worth it after thinking more about it

ktoso commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedSetup.scala:
```diff
@@ -0,0 +1,67 @@
+  recovery:          Recovery,
+
+  settings: EventsourcedSettings,
+
```
ok will remove, they helped me a bit to see what is "grouped" conceptually

ktoso commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/PersistentBehaviors.scala:
```diff
@@ -137,19 +67,101 @@ object PersistentBehaviors {
+}
+
+@InternalApi
+private[akka] final case class PersistentBehaviorImpl[Command, Event, State](
```
good point!

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala:
```diff
@@ -0,0 +1,325 @@
+            if (setup.snapshotWhen(newState2.state, event, newState2.seqNr))
+              internalSaveSnapshot(newState2)
+          }
+          val pendingInvocations = StashingHandlerInvocation(event, handler) :: Nil
```
Don't use that untyped `StashingHandlerInvocation` class here.
I think this can be simplified. We only support one `Persist` or `PersistAll` effect, right?
So we only need one callback, not a list.
And we don't care about the event.
Meaning that `pendingInvocations` can be `afterPersist: () => Unit`

patriknw replied:
cont a few lines below...

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala:
```diff
@@ -0,0 +1,325 @@
+
+            val pendingInvocations = events.map { event ⇒
+              StashingHandlerInvocation(event, handler)
+            }
```
ok, here we have more than one callback, but we don't really need this pendingInvocations stack here. We can have an event counter and do the stuff when that many `WriteMessageSuccess` have been received.
We actually don't need the callback at all, since we only use it for snapshotting.
Meaning that we can replace `pendingInvocations` with `numberOfEvents: Int` and `snapshotAfterPersist: Boolean`.

Let me know if these comments are difficult to follow, and I can do that refactoring instead.

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala:
```diff
@@ -0,0 +1,325 @@
+
+    final def onJournalResponse(
+      response: Response): Behavior[InternalProtocol] = {
+      setup.log.debug("Received Journal response: {}", response)
```
not sure we dare `toString` of those messages?

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala:
```diff
@@ -0,0 +1,325 @@
+      setup.log.debug("Received Journal response: {}", response)
+      response match {
+        case WriteMessageSuccess(p, id) ⇒
+          // instanceId mismatch can happen for persistAsync and defer in case of actor restart
```
I don't think that is specific for persistAsync

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala:
```diff
@@ -0,0 +1,325 @@
+          // instanceId mismatch can happen for persistAsync and defer in case of actor restart
+          // while message is in flight, in that case the handler has already been discarded
+          if (id == setup.writerIdentity.instanceId) {
+            // onWriteMessageComplete() -> tryBecomeHandlingCommands
```
remove ?

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala:
```diff
@@ -0,0 +1,325 @@
+    }
+
+    //    private def onWriteMessageComplete(): Unit =
+    //      tryBecomeHandlingCommands()
```
remove

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala:
```diff
@@ -0,0 +1,325 @@
+    private def onSnapshotterResponse(response: SnapshotProtocol.Response): Behavior[InternalProtocol] = {
+      response match {
+        case SaveSnapshotSuccess(meta) ⇒
+          setup.context.log.debug("Save snapshot successful: " + meta)
```
use log placeholders

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala:
```diff
@@ -0,0 +1,325 @@
+          setup.context.log.debug("Save snapshot successful: " + meta)
+          this
+        case SaveSnapshotFailure(meta, ex) ⇒
+          setup.context.log.error(ex, "Save snapshot failed! " + meta)
```
^

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala:
```diff
@@ -0,0 +1,325 @@
+    // manual loop implementation to avoid allocations and multiple scans
+    while (it.hasNext) {
+      val effect = it.next()
+      applySideEffect(effect, state) match {
```
```
if (!applySideEffect(effect, state).isAlive)
  res = Behaviors.stopped
```

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala:
```diff
@@ -0,0 +1,325 @@
+  }
+
+  def applySideEffect(effect: ChainableEffect[_, S], state: EventsourcedState[S]): Behavior[InternalProtocol] = effect match {
+    case _: Stop.type @unchecked ⇒
```
`case Stop ⇒` ?

patriknw reviewed (commented):
Review comments for EventsourcedRecoveringEvents and EventsourcedRunning. That concludes my detailed review.

Summary:
* Returning of permits must be more robust
* pendingInvocations can be simplified
* Many many small simple fixes
* More test should be added, but that can be done in separate PR. Created issue https://github.com/akka/akka/issues/24687

ktoso commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedSettings.scala:
```diff
@@ -0,0 +1,92 @@
+
+    val stashCapacity = typedConfig.getInt("stash-capacity")
+
+    val stashingLogLevel = typedConfig.getString("log-stashing") match {
```
yup

ktoso commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedSettings.scala:
```diff
@@ -0,0 +1,92 @@
+
+    // StashOverflowStrategy
+    val internalStashOverflowStrategy =
+      untypedConfig.getString("internal-stash-overflow-strategy") // FIXME or copy it to typed?
```
good

ktoso commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedSetup.scala:
```diff
@@ -0,0 +1,67 @@
+  def selfUntyped = context.self.toUntyped
+
+  import EventsourcedBehavior.InternalProtocol
+  val selfUntypedAdapted: a.ActorRef = context.messageAdapter[Any] {
```
excellent :)

ktoso commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedStashManagement.scala:
```diff
@@ -0,0 +1,57 @@
+    }
+  }
+
+  // FIXME, yet we need to also stash not-commands, due to journal responses ...
```
Hm, my mistake I think, we never stash journal responses, all good

ktoso commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedStashManagement.scala:
```diff
@@ -0,0 +1,57 @@
+  protected def tryUnstash(
+    behavior: Behavior[InternalProtocol]): Behavior[InternalProtocol] = {
+    if (stashBuffer.nonEmpty) {
+      setup.log.debug("Unstashing message: {}", stashBuffer.head.getClass)
```
done

ktoso reviewed (commented):
Addressed all comments, tests passing; Adding more tests as separate ticket

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRecoveringEvents.scala:
```diff
@@ -0,0 +1,185 @@
+      }
+    } else {
+      // snapshot timeout, but we're already in the events recovery phase
+      Behavior.unhandled
```
That's the nice thing with timers, once you canceled it will not be received even though it's already enqueued in mailbox. It's ok to keep the unhandled for this. I just wanted to point out that it shouldn't happen.

ktoso replied:
Ah excellent, I had forgotten about this :)

patriknw commented on akka-persistence-typed/src/main/resources/reference.conf:
```diff
@@ -0,0 +1,15 @@
+  # enables automatic logging of messages stashed automatically by an PersistentBehavior,
+  # this may happen while it receives commands while it is recovering events or while it is persisting events
+  # Set to a log-level (debug, info, warn, error) to log stashed messages on given log level;
+  log-stashing = off
```
We don't have such in other places. It's typical tracing/debugging logging so it makes most sense to log it at debug level. Especially since it `.toString` each message, which is something I don't want anywhere apart from verbose debug.

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedJournalInteractions.scala:
```diff
@@ -73,16 +70,41 @@ private[akka] trait EventsourcedJournalInteractions[C, E, S] {
   }
 
+  /** Mutates setup, by setting the `holdingRecoveryPermit` to false */
+  protected def tryReturnRecoveryPermit(reason: String): Unit = {
```
this is called from `onRecoveryFailure`, which shouldn't be necessary since it's done in PreRestart and PostStop.

It must still be done from `onRecoveryCompleted` ofc.

ktoso replied:
Yes though there it can include beffer information about the reason: ` tryReturnRecoveryPermit("on replay failure: " + cause.getMessage)` o to keep? Due to the flag it's not incorrect, it will be a no-op in PostStop

patriknw replied:
ok

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRecoveringSnapshot.scala:
```diff
@@ -79,87 +68,61 @@ private[akka] class EventsourcedRecoveringSnapshot[C, E, S](
-      returnRecoveryPermitOnlyOnFailure(cause)
-      throw cause
+  def onJournalResponse(response: JournalProtocol.Response): Behavior[InternalProtocol] = {
+    setup.log.debug("Unexpected response from journal: [{}], may be due to an actor restart, ignoring...", response)
```
Not sure we dare `toString` on those messages, class name should be enough and safe.

patriknw replied:
and yes, people use debug level in production

ktoso replied:
good, agreed -- just getName here

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRecoveringEvents.scala:
```diff
@@ -175,8 +171,8 @@ private[persistence] class EventsourcedRecoveringEvents[C, E, S](override val se
+  // protect against event recovery stalling forever because of journal overloaded and such
+  private val RecoveryTickTimerKey = "event-recovery-tick"
   private def startRecoveryTimer(timers: TimerScheduler[InternalProtocol], timeout: FiniteDuration): Unit =
     timers.startPeriodicTimer(RecoveryTickTimerKey, RecoveryTickEvent(snapshot = false), timeout)
```
`snapshot = true`, but I don't think the flag is needed at all

ktoso replied:
Yeah with the timer cancelations I see, I'll keep it for now though since timer semantics may change -- see Johan's PR

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRecoveringSnapshot.scala:
```diff
@@ -38,33 +36,24 @@ private[akka] object EventsourcedRecoveringSnapshot {
         case (_, RecoveryTickEvent(snapshot)) ⇒ onRecoveryTick(snapshot)
         case (_, cmd: IncomingCommand[C])     ⇒ onCommand(cmd)
-      }
+      }.onSignal(returnPermitOnStop)
```
nice

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala:
```diff
@@ -161,12 +164,12 @@ import scala.collection.immutable
-      }
+        case (_, SnapshotterResponse(_))           ⇒ Behaviors.unhandled
+        case (_, JournalResponse(_))               ⇒ Behaviors.unhandled
+      }.onSignal(returnPermitOnStop)
```
In Running this shouldn't be needed because the permit is always returned before that phase.

ktoso replied:
right, thanks

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala:
```diff
@@ -197,6 +200,9 @@ import scala.collection.immutable
     }
 
+    override def onSignal: PartialFunction[Signal, Behavior[InternalProtocol]] =
+      { case signal ⇒ returnPermitOnStop((setup.context, signal)) }
```
not needed

ktoso replied:
Right thanks; This exposed potential need for docs in case someone wants to share such handler which I added as todo to the guide https://github.com/akka/akka/issues/22805 :)

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala:
```diff
@@ -301,10 +288,8 @@ import scala.collection.immutable
-        case _                     ⇒ // nothing to do
-      }
+      val stopped = !Behavior.isAlive(applySideEffect(effect, state))
+      if (stopped) res = Behaviors.stopped
```
👍

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedStashManagement.scala:
```diff
@@ -17,23 +20,25 @@ private[akka] trait EventsourcedStashManagement[C, E, S] {
-    if (logLevel != Logging.OffLevel) context.log.debug("Stashing message: {}", msg) // FIXME can be log(logLevel once missing method added
-
-    val internalStashOverflowStrategy: StashOverflowStrategy = setup.persistence.defaultInternalStashOverflowStrategy
+    val internalStashOverflowStrategy: StashOverflowStrategy = {
```
shouldn't this be done once, instead of for each message?
perhaps even put it in `setup`

patriknw replied:
not fixed yet

ktoso replied:
Missed this, thanks!

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala:
```diff
@@ -3,39 +3,40 @@
  */
-@InternalApi private[akka] object EventsourcedRunning {
+@InternalApi
+private[akka] object EventsourcedRunning {
```
the simplifications of `pendingInvocations` are not done, I can take a stab at that if you like

patriknw replied:
Here is the simplification of the pendingInvocations: fc5e5cfbd1079c1ffec7d57004c8c5496b3d7758
Probably best to apply that before other changes.

patriknw reviewed (commented):
Reviewed the changes after previous review.
Let me know if I missed any questions in the outdated comments.

patriknw committed:
simplify pendingInvocations

patriknw commented:
Here is the simplification of the pendingInvocations: fc5e5cfbd1079c1ffec7d57004c8c5496b3d7758
Probably best to apply that before other changes.

ktoso commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedJournalInteractions.scala:
```diff
@@ -73,16 +70,41 @@ private[akka] trait EventsourcedJournalInteractions[C, E, S] {
+      tryReturnRecoveryPermit("PostStop")
+      Behaviors.stopped
+    case (_, PreRestart) ⇒
+      // TODO was not entirely sure if it's needed here as well
```
Got it, good to document in typed patterns

ktoso commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedJournalInteractions.scala:
```diff
@@ -73,16 +70,41 @@ private[akka] trait EventsourcedJournalInteractions[C, E, S] {
+    } // else, no need to return the permit
+  }
+
+  protected def returnRecoveryPermitOnlyOnFailure(cause: Throwable): Unit =
```
thanks

ktoso committed:
recovering = replaying naming change

make all matches exhaustive

internalStashOverflowStrategy should be initialized once

ktoso commented:
Thanks, addressed comments, made matches exhaustive and included your commit.

ktoso commented:
May be good to look at https://github.com/akka/akka/pull/24681/commits/a89eff15642b54326db8fd5b5fb57efd95334162 in case we want to handle those differently

ktoso commented:
And a ticket for handling deletes

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRecoveringEvents.scala:
```diff
@@ -0,0 +1,182 @@
+}
+
+@InternalApi
+private[persistence] class EventsourcedRecoveringEvents[C, E, S](override val setup: EventsourcedSetup[C, E, S])
```
You keep ignoring my comments about the naming here?
Terminology:
recovery = snapshot + replay

Meaning that this should be named EventsourcedReplayingEvents (and probably several more things in this file recovering -> replaying)

ktoso replied:
Sorry, I misunderstood your comment about it I think. I'm fine with renaming to Replaying

ktoso replied:
Thought it was only about log statements for some reason which I changed

ktoso replied:
Addressed now, sorry, ignoring was not intentional

patriknw commented:
noticed 2 review comments that were not addressed

ktoso commented:
- strategy moved to setup
- recovering => replaying everywhere

patriknw reviewed (approved):
LGTM, yeah it was a lot of iterations but the end result is great and we learnt several things on the way

ktoso commented:
Uff, exciting! Will squash a bit but keep authorship of people who helped out in-tact and merge then.

ktoso commented:
Thanks for the tireless and excellent reviews and help in pushing this over the line.
Some minor pending issues are ticketified and can be taken one by one from here.

ktoso commented:
Resolves  #23694
